Address safer safer C++ warnings for SourceBuffer.h

eric-carlson opened the pull request:
#### 02869d6b93d6715c614b01190516e4e4ec41cf33
<pre>
Address safer safer C++ warnings for SourceBuffer.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=287381">https://bugs.webkit.org/show_bug.cgi?id=287381</a>
<a href="https://rdar.apple.com/144498751">rdar://144498751</a>

Reviewed by Jean-Yves Avenard.

* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::setTimestampOffset):
(WebCore::SourceBuffer::abort):
(WebCore::SourceBuffer::remove):
(WebCore::SourceBuffer::rangeRemoval):
(WebCore::SourceBuffer::changeType):
(WebCore::SourceBuffer::virtualHasPendingActivity const):
(WebCore::SourceBuffer::appendBufferInternal):
(WebCore::SourceBuffer::sourceBufferPrivateAppendComplete):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveRenderingError):
(WebCore::SourceBuffer::setActive):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
(WebCore::SourceBuffer::appendError):
(WebCore::SourceBuffer::sourceBufferPrivateDurationChanged):
(WebCore::SourceBuffer::sourceBufferPrivateDidDropSample):
(WebCore::SourceBuffer::setMode):
(WebCore::SourceBuffer::updateBuffered):
(WebCore::SourceBuffer::setBufferedDirty):
(WebCore::SourceBuffer::memoryPressure):
(WebCore::SourceBuffer::sourceBufferPrivateDidAttach):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/290166@main">https://commits.webkit.org/290166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3452bd0c2a5ce044d9e9ac815c2ef5135da963ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8643 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43937 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94098 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39880 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9043 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16831 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68664 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26336 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92121 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6927 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80877 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49025 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6676 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35256 "Found 60 new test failures: compositing/animation/repaint-after-clearing-shared-backing.html compositing/repaint/iframes/compositing-iframe-scroll-repaint.html fast/repaint/align-self-change.html fast/repaint/align-self-overflow-change.html fast/repaint/justify-items-legacy-change.html fast/repaint/justify-self-change.html fast/repaint/justify-self-overflow-change.html fonts/monospace.html fonts/sans-serif.html fonts/serif.html ... (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38987 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77039 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36241 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95932 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16299 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11937 "Found 5 new test failures: http/tests/webarchive/cross-origin-stylesheet-crash.html http/tests/webarchive/test-css-url-encoding-shift-jis.html http/tests/webarchive/test-css-url-encoding-utf-8.html http/tests/webarchive/test-css-url-encoding.html http/tests/webarchive/test-preload-resources.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77544 "Failure limit exceed. At least found 1 new test failure: svg/batik/text/smallFonts.svg (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16555 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76667 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76835 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18950 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21239 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19848 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9391 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16313 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21624 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16054 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19505 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17835 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -233,7 +233,7 @@ ExceptionOr<void> SourceBuffer::setTimestampOffset(double offset)
     // 2. If this object has been removed from the sourceBuffers attribute of the parent media source, then throw an
     //    InvalidStateError exception and abort these steps.
     // 3. If the updating attribute equals true, then throw an InvalidStateError exception and abort these steps.
-    if (isRemoved() || m_updating)
```
`isRemoved()` checks that m_source is not null already

jyavenard commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -452,14 +455,20 @@ void SourceBuffer::rangeRemoval(const MediaTime& start, const MediaTime& end)
 
     // 5. Return control to the caller and run the rest of the steps asynchronously.
-    m_client->ensureWeakOnDispatcher([producer = WTFMove(producer), this, start, end](SourceBuffer&) mutable {
+    m_client->ensureWeakOnDispatcher([producer = WTFMove(producer), weakThis = WeakPtr { *this }, this, start, end](SourceBuffer&) mutable {
```
that's not needed, this is what `ensureWeakOnDispatcher` does 
```
            if (RefPtr parent = weakParent.get(); parent && !parent->isRemoved())
                function(*parent);
```

it checks if `isRemoved()` which checks for m_source

jyavenard commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -434,7 +437,7 @@ void SourceBuffer::rangeRemoval(const MediaTime& start, const MediaTime& end)
     MediaPromise::AutoRejectProducer producer(PlatformMediaError::BufferRemoved);
     scriptExecutionContext()->enqueueTaskWhenSettled(producer.promise(), TaskSource::MediaElement, [weakThis = WeakPtr { *this }, this](auto&&) {
-        if (!weakThis || isRemoved())
+        if (!weakThis || !m_source || isRemoved())
```
redundant, not needed

jyavenard commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -392,7 +392,7 @@ ExceptionOr<void> SourceBuffer::remove(const MediaTime& start, const MediaTime&
     //    an InvalidStateError exception and abort these steps.
     // 2. If the updating attribute equals true, then throw an InvalidStateError exception and abort these steps.
-    if (isRemoved() || m_updating)
+    if (!m_source || isRemoved() || m_updating)
```
not needed

jyavenard commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -590,6 +599,9 @@ void SourceBuffer::scheduleEvent(const AtomString& eventName)
 
 ExceptionOr<void> SourceBuffer::appendBufferInternal(std::span<const uint8_t> data)
 {
+    if (!m_source)
```
not needed.
Just below it does that already
```
    if (isRemoved() || m_updating)
        return Exception { ExceptionCode::InvalidStateError };
```

jyavenard commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -661,6 +673,9 @@ void SourceBuffer::sourceBufferPrivateAppendComplete(MediaPromise::Result&& resu
     if (isRemoved())
         return;
 
+    if (!m_source)
```
redundant

jyavenard commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -704,7 +719,8 @@ void SourceBuffer::sourceBufferPrivateDidReceiveRenderingError(int64_t error)
     ERROR_LOG(LOGIDENTIFIER, error);
 
-    m_source->streamEndedWithError(MediaSource::EndOfStreamError::Decode);
+    if (!!m_source)
```
`if (!isRemoved())`

jyavenard commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -773,14 +789,18 @@ void SourceBuffer::setActive(bool active)
 {
     ALWAYS_LOG(LOGIDENTIFIER);
 
+    if (!m_source)
```
This can't happen, so an assert would be sufficient. but I guess better safe than sorry

eric-carlson committed:
Address safer safer C++ warnings for SourceBuffer.h
https://bugs.webkit.org/show_bug.cgi?id=287381
rdar://144498751

Reviewed by Jean-Yves Avenard.

* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::setTimestampOffset):
(WebCore::SourceBuffer::abort):
(WebCore::SourceBuffer::remove):
(WebCore::SourceBuffer::rangeRemoval):
(WebCore::SourceBuffer::changeType):
(WebCore::SourceBuffer::virtualHasPendingActivity const):
(WebCore::SourceBuffer::appendBufferInternal):
(WebCore::SourceBuffer::sourceBufferPrivateAppendComplete):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveRenderingError):
(WebCore::SourceBuffer::setActive):
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
(WebCore::SourceBuffer::appendError):
(WebCore::SourceBuffer::sourceBufferPrivateDurationChanged):
(WebCore::SourceBuffer::sourceBufferPrivateDidDropSample):
(WebCore::SourceBuffer::setMode):
(WebCore::SourceBuffer::updateBuffered):
(WebCore::SourceBuffer::setBufferedDirty):
(WebCore::SourceBuffer::memoryPressure):
(WebCore::SourceBuffer::sourceBufferPrivateDidAttach):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:

Canonical link: https://commits.webkit.org/290166@main
